### ISSUE TITLE
incomplete example in Technical Notes > Error Handling

### ISSUE THREAD

**Issue by lucaferranti** [chapel-lang/chapel#20802]
The [first example](https://chapel-lang.org/docs/technotes/errorHandling.html?highlight=canthrow#fatal-mode-for-implicit-and-prototype-modules) in Technical Notes > Error Handling says

```chapel
canThrow(1); // handling can be omitted; halts if an error occurs

proc throwsErrorsOn() throws {
  // error propagates out of this function
  canThrow(-1);
}

proc doesNotThrowErrorsOn() {
  // causes a halt if called
  alwaysThrows();
}
```

but the function `canThrow` is not defined anywhere.

A similar running example is used in the [language specifications](https://chapel-lang.org/docs/language/spec/error-handling.html) , where `canThrow` is defined.

Would it make sense to merge the two error handling sections (while keeping the one in primer as more lightweight introduction)? From what I can say, they significantly overlap already

**Comment by bradcray** [comment]
Generally speaking, technical notes are used to describe features when they're new enough / evolving enough that we're not certain we're ready for them to be in the language specification yet.  Once a feature has made it into the language specification, we typically retire its technical note.  In this case, I don't happen to recall whether there is a reason that we're keeping the technical note around.  Looking at the PR message that incorporated it into the spec in https://github.com/chapel-lang/chapel/pull/14025, it sounds as though perhaps there isn't, although commit https://github.com/chapel-lang/chapel/pull/14025/commits/5de87842fba83697ca469f5267757402f4eae65c removed some duplication, but obviously not all of it.  Let's check with @mppf who did the work to see what he remembers.

My interpretation of the technical note is that it's not intended to be complete / executable code per se, but the routine named `canThrow()` is meant to be understood as a routine that potentially throws without needing to see its definition.  That's not to say that it couldn't/shouldn't define `canThrow()` as well, but if I were reading it, I don't know that it would've occurred to me that it was problematic that it did not (similar to how when I see an example saying `myClassVar.foo()` I understand that `myClassVar` is a class variable without necessarily needing to see it defined).

**Comment by mppf** [comment]
@bradcray - the main thing that I am aware of that is missing from the spec but present in the technote is discussion of error handling modes. We have a "strict mode" which can only be activated by a pragma and also the "fatal mode" which you can get with `prototype module`. We could consider deprecating these or documenting them in the spec. (and IMO `prototype` would be better represented as an attribute like `@prototype` etc but I know I might not be with the majority on such a use of attributes).

**Comment by lucaferranti** [comment]
@bradcray after reading your message, the example makes indeed more sense. Funnily, if I had indeed read something like `myClassVar.foo()` I would have interpreted as a placeholder, but for some strange reason, my brain overthought the line

```chapel
canThrow(1); // handling can be omitted; halts if an error occurs
```

and interpreted as `canThrow` being some sort of internal function to set a global state on how to handle exceptions 🤣 

The intro to technical notes is pretty clear

> Technical Notes describe in-progress language features that are not yet stable enough to be in the language specification. Additionally, a Technical Note can describe implementation details that can be ignored by most Chapel programmers.

I missed that because I landed on that page via the search engine. This was more on me that on the docs :) I think the issue can be closed

**Comment by bradcray** [comment]
> but for some strange reason my brain overthought the line

Arguably, a name like `myThrowingFn()` might have been more clear...

To Michael's comment:

Ah yes, those error-handling modes.  I think it'd be interesting to open up a new issue asking whether we can deprecate and get rid of them.  I have to admit, I don't think I've thought of them in years, and also haven't been unhappy with the current mode.  The prototype vs. production  module question is more interesting and at the moment I like the idea of replacing the `prototype` keyword with an annotation/attribute.  Will stew on that one some more...

I've captured these thoughts in https://github.com/chapel-lang/chapel/issues/20856 and am going to close this one since Luca's primary concerns have been addressed, and I think the salient points of the discussion are now captured there.
